src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java

```diff
@@ -0,0 +1,213 @@
+package frc.team2412.robot.subsystems;
+
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+import edu.wpi.first.wpilibj.shuffleboard.ShuffleboardTab;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.team2412.robot.commands.drivebase.DriveCommand;
+import java.util.function.DoubleSupplier;
+
+public class LimelightSubsystem extends SubsystemBase {
+
+	// CONSTANTS
+
+	private static final PIDController TRANSLATION_PID = new PIDController(10.0, 0, 0);
+	private static final PIDController ROTATION_PID = new PIDController(8.0, 0, 0);
+
+	// meters?
+	public static final double CAMERA_MOUNT_HEIGHT = 0.1143;
+	public static final double CAMERA_ANGLE_OFFSET = 0;
+	public static final double TARGET_HEIGHT = 0.33;
+
+	public static final double GOAL_DISTANCE_FROM_TARGET = 0.7;
+	public static final double GOAL_DISTANCE_FROM_CONE = 0.3;
+	public static final double GOAL_DISTANCE_FROM_CUBE = 0.3;
+
+	// MEMBERS
+
+	NetworkTable networkTable;
+
+	String currentPoseString;
+	String targetPoseString;
+
+	// network tables
+
+	// CONSTRUCTOR !
+	public LimelightSubsystem() {
+
+		// broadcast 10.24.12.227
+
+		// logging
+		currentPoseString = "";
+		targetPoseString = "";
+
+		networkTable = NetworkTableInstance.getDefault().getTable("limelight");
+		ShuffleboardTab limelightTab = Shuffleboard.getTab("Limelight");
+
+		setPipeline();
+
+		limelightTab.addBoolean("hasTarget", this::hasTargets).withPosition(0, 0).withSize(1, 1);
+		limelightTab
+				.addDouble("Horizontal Offset", this::getHorizontalOffset)
+				.withPosition(1, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addDouble("Vertical Offset", this::getVerticalOffset)
+				.withPosition(2, 0)
+				.withSize(1, 1);
+
+		limelightTab
+				.addDouble("Target Distance ", this::getDistanceFromTarget)
+				.withPosition(3, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addDouble("Target Distance 2 - TEST ", this::getDistanceFromTargetTheSecond)
+				.withPosition(4, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addDouble("Limelight Based Turn Power - TEST ", this::turnPowerLin)
+				.withPosition(5, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addString("Current Pose ", this::getCurrentPoseString)
+				.withPosition(0, 1)
+				.withSize(4, 1);
+
+		limelightTab
+				.addString("Target Pose ", this::getTargetPoseString)
+				.withPosition(0, 2)
+				.withSize(4, 1);
+	}
+
+	public void setPipeline() {
+		networkTable.getEntry("pipeline").setNumber(2);
```

Comment:
Can be `private` as nothing outside the class calls this.

Reply:
done